Implement Lagrange polynomial interpolation

WebFreak001 opened the pull request:
This PR adds Lagrange polynomial calculation functions into the MathUtils namespace.

Polynomial interpolations are useful for example for giving the user the ability to specify curves using simple arrays of double values. A typical use-case is for example the "Curves" tool in photo editing software which uses polynomial interpolations to interpolate between the points a user gave.

I'm hoping this could be useful inside osu for example in mapping for being able to provide customizable curve functions for animatable values like for example ppy/osu#3953. It's not quite as customizable as being able to provide custom math functions anywhere, but it's a lot more space-efficient just storing a few coordinate pairs to describe a close approximation to any "curvy" function.

The advantage of this over bezier curves is that it is a mathematical _function_, so any one input has exactly one output.

![Polynomial interpolation](http://mathworld.wolfram.com/images/eps-gif/LagrangeInterpolatingPoly_900.gif)

Tom94 commented:
Hey there, nice addition! My main gripe with polynomial interpolation is the diverging behavior as the number of points becomes large ([wikipedia](https://en.wikipedia.org/wiki/Lagrange_polynomial) has a nice illustration of this; also known as Runge's phenomenon).

It would be awesome if you could add an additional function with restricted interface that enforces the x-coordinates to follow [Chebyshev nodes](https://en.wikipedia.org/wiki/Chebyshev_nodes), minimizing this divergence phenomenon.

WebFreak001 commented:
I think Chebyshev nodes are a little bit out of scope of this PR as the polynomial interpolation is intended to be used for users to make curves (who will adjust the curve if they see that it becomes really weird, just like in Photoshop for example), rather than interpolating measured data points for now. I think instead of using Lagrange polynomials for such automatic data interpolation, other algorithms like using splines should be preferred.

Tom94 commented:
Does it go away if you don't define a local variable and instead pass `null` directly?

smoogipoo commented:
This stuff would fit better inside `MathUtils.PathApproximator` no? Also the basis stuff should probably be private methods?

WebFreak001 commented:
hm Lagrange interpolation can also extrapolate further than the given points, which might not really be a proper use-case of PathApproximator.

Maybe an implementation both in PathApproximator and inside Interpolation could be useful for generating a full curve of the function from 0 to 1 for rendering inside UI for example. (It could also reduce the computation because if you repeatedly calculate over it you can optimize Lagrange)

I can make the basis polynomials private, they are some properly defined math functions used inside the polynomial but I don't know if they have any use anywhere else outside Lagrange.

WebFreak001 commented:
I added barycentric lagrange interpolation now which makes calculation O(n) when repeatedly calculating over the same lagrange polynomial. I also added a PathApproximator implementation with it and a visual test case to see the PathApproximators in action for some 5-point path. (though the arc only takes the first 3)

Chebyshev nodes should be implementable more easily with the PathApproximator imlpementation too (it currently just generates 51 equally spaced points)

smoogipoo commented:
I'm fine with getting this in in its current state and building upon it with Chebyshev nodes afterwards, if you'd like. Seems pretty solid for now.
